File: 프로그래머스/2/42860. 조이스틱/조이스틱.kt
import kotlin.math.abs
import kotlin.math.min

class Solution {
    fun countMove(name: String, currentIndex: Int, count: Int, trial: Int): Int {
        val newName = name.slice(0 until currentIndex) + 'A' + name.slice(currentIndex + 1 until name.length)
        val currentMoveCount = min(abs(name[currentIndex] - 'A'), abs(name[currentIndex] - 'Z') + 1)
        
        if (newName.all { it == 'A' }) return count + currentMoveCount
        
        var left = if (currentIndex == 0) name.length - 1 else currentIndex - 1
        var leftMoveCount = 1
        while (name[left] == 'A') {
            left = if (left == 0) name.length - 1 else left - 1
            leftMoveCount += 1
        }
        
        var right = (currentIndex + 1) % name.length
        var rightMoveCount = 1
        while (name[right] == 'A') {
            right = (right + 1) % name.length
            rightMoveCount += 1
        }
        
        
        return Math.min(countMove(newName, left, count + currentMoveCount + leftMoveCount, trial + 1), countMove(newName, right, count + currentMoveCount + rightMoveCount, trial + 1))
    }
    
    fun solution(name: String): Int {
        return countMove(name, 0, 0, 0)
    }
}


// 5 0
// 5 1 => 4
// 5 2 => 3
// 5 3 => 2
// 5 4 => 1
// 5 5 => 0
// 5 6 => -1 => 9
// 5 7 => -2 => 8
// 5 8 => -3 => 7
// 5 9 => -4 => 6
// 5 10 => -5 => 5
// 5 11 => -6 => 4
// 5 12 => -7 => 3
// 5 13 => -8 => 2
// 5 14 => -9 => 1
// 5 15 => -10 => 0
// 5 16 => -11 => 9


// length - (right % currentIndex)
// 10 - (6 % 5)
// 10 - (1 % 5) 

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
from math import inf
from collections import deque

def vertical_distance(char):
    start = ord('A')
    end = ord('Z')
    target = ord(char)

    return min(abs(target - start), abs(end - target + 1))

def calculate_directional_distances(n, current, next):
    normal_distance = next - current
    reversed_distance = (n - next) + current + 1
    return [normal_distance, reversed_distance]



def solution(name):
    count = 0
    for char in name:
        count += vertical_distance(char)

        
    
    positions = deque()
    for i in range(0, len(name)):
        char = name[i]
        if char != "A":
            positions.append(i)
            
    if len(positions) == 0:
        return count
    
    if len(positions) == 1:
        return count + min(calculate_directional_distances(len(name) - 1, 0, positions[0]))
    
    distance = inf
    
    normal = positions[len(positions) - 1]
    opposite = len(name) - positions[0]
    
    
    distance = min(distance, normal)
    distance = min(distance, opposite)
    
    for i in range(0, len(positions) - 1):
        reversed1 = 2 * positions[i] + len(name) - positions[i + 1]
        reversed2 = (len(name) - positions[i + 1]) * 2 + positions[i]
        distance = min(distance, reversed1)
        distance = min(distance, reversed2)
        
    
        
    print(distance, count)
            
    return distance + count
    

# 1. positions[i]를 기준으로 회전하여, positions[i + 1]에 도착 (정방향 / 역방향) (len(positions) > 2)
# 2. 처음부터 끝까지 정방향
# 3. 처음부터 역방향


    
        
        
        
        

    
        
    
    
    
    # 역방향 우선
    
    return count


# min(처음것 조작 + 역방향 or 정방향)
# 각 알파벳 도달: Z부터 이동 + 1 vs A부터 이동

# 반례 AZAAAAZ
# 다시 돌아오는 왕복식으로 하는 게 더 빠름
# 정방향 / 역방향으로 나눠서 항상 가장 가까운 놈에 도달하는 알고리즘 구성


